### ISSUE TITLE
W8003 False positive when parameters have default values

### ISSUE THREAD

**Issue by CarlyG55** [aws-cloudformation/cfn-lint#3899]
### CloudFormation Lint Version

1.22.3

### What operating system are you using?

Ubuntu

### Describe the bug

The linter errors with W8003 on the following example, since it thinks the value of the parameter is always the default value, when there are multiple possible options that would lead to true or false

### Expected behavior

W8003 is not raised

### Reproduction template

```yaml
Parameters:
  Environment:
    Description: "The name of the environment to deploy to."
    Type: "String"
    Default: local
    AllowedValues:
      - "local"
      - "staging"
      - "integration"
      - "production"
    DeploymentName:
    Type: String
    MaxLength: 22
    AllowedPattern: ^.*[^-]$
    Default: test
    Description: A unique prefix to identify the deployment; used to distinguish variation between ephemeral stacks

Rules:
  DeploymentNameRequired:
    RuleCondition: !Equals [ !Ref Environment, "local" ]
    Assertions:
      - Assert: !Not [ !Equals [ !Ref DeploymentName, "" ] ]
        AssertDescription: >
          Must specify DeploymentName parameter when Environment is "local"
```

**Comment by kddejong** [comment]
I'm not able to replicate this with the provided template.  How are you using these parameters?  Also do you have any conditions?

**Comment by CarlyG55** [comment]
Hi @kddejong, thanks for the quick response!
I can point you to the exact failure if that's helpful - I thought it would be helpful to simplify it for the example but maybe it has lost some context. It's a codebase I've recently started maintaining, and the warning has just starting appearing.

The failure can be found [here](https://github.com/govuk-one-login/onboarding-self-service-experience/actions/runs/12670806700/job/35311219675), although I don't think you can see the logs - the relevant output is
```
W8003 ['local', 'local'] will always return True or False
backend/api/api.template.yml:40:5

W8003 ['self-service', ''] will always return False
backend/api/api.template.yml:42:24

W8003 ['local', 'local'] will always return True or False
backend/api/api.template.yml:47:27

W8003 ['self-service', 'self-service'] will always return True or False
backend/api/api.template.yml:49:9

W8003 [True, True] will always return True or False
backend/api/api.template.yml:100:3

W8003 ['local', 'local'] will always return True or False
backend/api/api.template.yml:105:3
```

The file that is failing can be found [here](https://github.com/govuk-one-login/onboarding-self-service-experience/blob/main/backend/api/api.template.yml). Thank you!

**Comment by kddejong** [comment]
Thanks @CarlyG55.  I'll look into it today.  I was able to replicate the issue with that template

**Comment by kddejong** [comment]
A quick analysis has this something to do with the transforms.  This is the result of the PR to fix #3857.  Some people are relying on the language extensions to replace certain Ref values before sending to SAM as SAM can't always handle not having a value. 

I think the easy fix here is to only do Ref replacements when in the Resources section and not in Rules/Conditions.

**Comment by CarlyG55** [comment]
Thank you @kddejong!
